### ISSUE TITLE
Please add a way to configure a task that is called before the debugger extension is called

### ISSUE THREAD

**Issue by mickare** [microsoft/vscode#112668]
- VSCode Version: 1.52.0
- OS Version:  Manjaro Linux 20.2
- Downstream issue: https://github.com/microsoft/vscode-cpptools/issues/5060
- Related issues:
  - https://github.com/microsoft/vscode/issues/79932
  - https://github.com/microsoft/vscode/issues/95371

This issue is the result and combination of multiple other issues that I think are based on a design flaw in the launch & task system of VSCode. The major problem is the order how things work in the launch/debug system. In a `launch.json` configration the `envFile` field allows to load environment variables from a file to setup the debugger environment. The `preLaunchTask` field allows to run some scripts before a debug run.

The issue is that the execution of `preLaunchTask` and the loading of `envFile` are executed in the wrong order to be practiable. Or you could say the current system is not flexible enough. See @WardenGnaw response to https://github.com/microsoft/vscode-cpptools/issues/5060#issuecomment-597319058. 
### TLDR; 
Basically the `envFile` is loaded when the debugger is configured; and the `preLaunchTask` is executed afterwards. 
The `envFile` is not reloaded when restarting the debug session (green arrow).
Current Launch-Workflow :   `load env` >  *DebugConfiguration*[ `run task` > `debug` ]

### User story
As a user, I would expect that the `preLaunchTask` behaves like the descripton states "Task to run before debug session starts." In my understanding this means, that it is run before any code in the debugger is called. And I would also expect that the `envFile` is loaded **after** the `preLaunchTask` has finished.
Expected Launch-Workflow: `run task` > `load env` > *DebugConfiguration*[ `debug` ]

There are many use cases for this like the automatic setup of the debug environment. E.g. the automatic detection of a free port to run the roscore and automatically configure the ROS_MASTER_URI.
Suggestions like setting the environment variables outside of vscode is not practicable in a remote ssh workspace with the possibility of running applications in parallel like ros workspaces (and with multiple users).

### Steps to Reproduce:
1. Try to load a task-generated environment file in a launch configuration

Does this issue occur when all extensions are disabled?: Yes

### Possible solutions
- **Reorder Solution**: Change the order of `envFile` and `preLaunchTask` by re-configuring the Debugger before each run
    `run task` > `load env` > *DebugConfiguration*[ `debug` ]
- **Minimal Solution**: Add a new task field to the launch configuration (e.g. `preConfigTask`) that is executed before configuring the debugger (meaning before loading the `envFile`).
    `run task` > `load env` > *DebugConfiguration*[ `run task` > `debug`]

### Workarounds ?
@isidorn @weinand @alexr00 @jnickg
Or do I miss something? Are there any **automatic** workarounds?

**Comment by weinand** [comment]
Please note: VS Code does not provide any support for loading environment variables from an `envFile`.
`envFile` support is an implementation detail of the JavaScript/Node debug extension (and maybe a few other extensions).

Any errors around `envFile` reading are the responsibility of the extension implementing the `envFile` support.

**Comment by mickare** [comment]
Then I will rephrase my request to match the User-Story: Please add a way to configure a task that is called before the debugger extension is called.

**Comment by weinand** [comment]
@mickare what exactly do you mean by "add a way to configure a task"?

**Comment by weinand** [comment]
Please see my comment https://github.com/microsoft/vscode-cpptools/issues/5060#issuecomment-747339576

**Comment by mickare** [comment]
> @mickare what exactly do you mean by "add a way to configure a task"?

Please add a field (e.g. `setupTask`) in `launch.json` that can be used to declare a task (from `tasks.json`) that is called before the launch extension is called.
```json
{
    "version": "2.0.0",
    "configurations": [
    {
        "name": "Debug",
        "type": "....",
        "preLaunchTask": " ... ",
        "setupTask": " .... "
    }
    ]
}
```

This could cover the general case of automatic setups and workflows while not having to rely on implementation details of extensions.

> Please see my comment [microsoft/vscode-cpptools#5060 (comment)](https://github.com/microsoft/vscode-cpptools/issues/5060#issuecomment-747339576)

Thanks.

**Comment by RandomGHUser** [comment]
I too would like to see this added.  I need to be able to run a task to build my environment variables before the debugger kicks in and grabs the envFile.  

Simple example, I have these environment files (Shared, Secrets, Development, Staging Production).  Before launching (or re-launching), I want to run a batch file to combine the relevant fields for "Development" (combining fields from the Shared, Development, and Secrets env files) into a single envFile, and then launch the program with that generated envFile.

**Comment by weinand** [comment]
The "preLaunchTask" **is** called before the debug session is started.
The problem you are seeing comes from the fact that cpptools has decided to load environment variables **before** the "preLaunchTask" is run. This is a specific cpptools issue. The JavaScript debug extensions load environment variables **after** the "preLaunchTask" is run.

We will not add a "setupTask" because we do not see a need for it.

**Comment by weinand** [comment]
With https://github.com/microsoft/vscode-cpptools/issues/5060#issuecomment-747691133 my statement from [above](https://github.com/microsoft/vscode-cpptools/issues/5060#issuecomment-747691133) was corrected: cpptools is loading the envFile after the "preLaunchTask" has been run. So the debuggee should see the updated environment variables.

@mickare can you confirm that this is the case?

**Comment by mickare** [comment]
> @mickare can you confirm that this is the case?

I can confirm that it works partially (using the cpptool). When the debugger is started for the first time it is loaded correctly.

![image](https://user-images.githubusercontent.com/2905154/102679048-dba2ec00-41ac-11eb-80c1-68c4aea3f64c.png)

But when restarting the debugger (Ctrl+Shift+F5)  the file is not reloaded, altough the `preLaunchTask` is executed.

> The envFile is not reloaded when restarting the debug session (green arrow).

![image](https://user-images.githubusercontent.com/2905154/102679111-77ccf300-41ad-11eb-93c8-d6065421ba72.png)

[Example Project (re-used from another issue)](https://github.com/mickare/vscode-broken-ssh-cpptools/tree/task)

**Comment by weinand** [comment]
@mickare 
> I can confirm that it works partially (using the cpptool). When the debugger is started for the first time it is loaded.

thanks for confirming that VS Code calls the extension API hooks in the correct order.

> But when restarting the debugger (Ctrl+Shift+F5) the file is not reloaded, altough the preLaunchTask is executed.

As said earlier, loading envFile is an implementation detail of cpptools and not part of the VS Code architecture.
I suggest that you comment on https://github.com/microsoft/vscode-cpptools/issues/5060 and ask them to reload the envFile on debug session restart.

/cc @WardenGnaw

**Comment by WardenGnaw** [comment]
@weinand Is `resolveDebugConfigurationWithSubstitutedVariables` not called on restart debugging? If not, shouldn't VS Code send the same launch config on restart to the debug adapter that it did for the first debug session?

**Comment by weinand** [comment]
@isidorn could you please answer Andrew's question?

**Comment by isidorn** [comment]
@weinand correct.
VS Code will call the `resolveDebugConfiguration` hooks again on restart only if the underlying configuration got changed by the user.

**Comment by weinand** [comment]
@isidorn thanks. That behavior makes sense.
